File: apps/api/plane/utils/github_app.py
"""
GitHub App authentication helpers.

Usage:
    from plane.utils.github_app import get_installation_access_token
    token = get_installation_access_token(installation_id)
    # Use token in Authorization: Bearer {token} header
"""
import base64
import binascii
import os
import time
from typing import Optional, Tuple

import requests
from plane.license.utils.instance_value import get_configuration_value


def _get_config_value(key: str) -> Optional[str]:
    """Read a value using the same InstanceConfiguration/env resolution as the rest of the app."""
    try:
        (value,) = get_configuration_value([{"key": key, "default": os.environ.get(key)}])
        return value
    except Exception:
        return os.environ.get(key)


def get_installation_access_token_result(installation_id: str) -> Tuple[Optional[str], Optional[str]]:
    """
    Return the installation access token plus a machine-readable error string.

    Error values:
      - github_app_not_configured
      - invalid_github_app_private_key
      - github_app_token_request_failed
    """
    app_id = _get_config_value("GITHUB_APP_ID")
    private_key_b64 = _get_config_value("GITHUB_APP_PRIVATE_KEY")

    if not app_id or not private_key_b64:
        return None, "github_app_not_configured"

    try:
        import jwt
        from cryptography.hazmat.primitives import serialization

        pem = base64.b64decode(private_key_b64)
        private_key = serialization.load_pem_private_key(pem, password=None)

        now = int(time.time())
        payload = {"iat": now - 60, "exp": now + 600, "iss": app_id}
        app_jwt = jwt.encode(payload, private_key, algorithm="RS256")
    except (ValueError, TypeError, binascii.Error):
        return None, "invalid_github_app_private_key"
    except Exception:
        return None, "github_app_token_request_failed"

    try:
        resp = requests.post(
            f"https://api.github.com/app/installations/{installation_id}/access_tokens",
            headers={
                "Authorization": f"Bearer {app_jwt}",
                "Accept": "application/vnd.github+json",
                "X-GitHub-Api-Version": "2022-11-28",
            },
            timeout=10,
        )
        if resp.ok:
            return resp.json().get("token"), None
        return None, "github_app_token_request_failed"
    except Exception:
        return None, "github_app_token_request_failed"


def get_installation_access_token(installation_id: str) -> Optional[str]:
    """
    Exchange a GitHub App installation_id for a short-lived installation access token.
    Reads GITHUB_APP_ID and GITHUB_APP_PRIVATE_KEY from InstanceConfiguration DB
    (set via God Mode) with fallback to environment variables.
    Returns None if credentials are not configured or the GitHub API call fails.
    """
    token, _ = get_installation_access_token_result(installation_id)
    return token

File: apps/api_rust/src/utils/github_app.rs
// src/utils/github_app.rs
//! Utilities for GitHub App authentication (JWT RS256 + installation access token).

use anyhow::Context;
use base64::{engine::general_purpose, Engine as _};
use jsonwebtoken::{encode, Algorithm, EncodingKey, Header};
use serde::{Deserialize, Serialize};

use crate::{error::AppError, utils::instance_config::get_instance_config, AppState};

#[derive(Debug, Serialize, Deserialize)]
struct AppClaims {
    /// Issued at — 60 s in the past to compensate for clock skew
    iat: i64,
    /// Expiry — maximum 10 min, we use 9 min for margin
    exp: i64,
    /// Issuer — GitHub App ID (numeric string)
    iss: String,
}

/// Generates a JWT RS256 signed with the GitHub App's private key.
///
/// Returns `Ok(None)` if the app is not configured (GITHUB_APP_ID /
/// GITHUB_APP_PRIVATE_KEY absent) to allow soft degradation without fatal error.
async fn build_app_jwt(state: &AppState) -> Result<Option<String>, AppError> {
    let app_id = get_instance_config(state, "GITHUB_APP_ID").await?;
    let key_b64 = get_instance_config(state, "GITHUB_APP_PRIVATE_KEY").await?;

    let (Some(app_id), Some(key_b64)) = (app_id, key_b64) else {
        return Ok(None); // GitHub App not configured
    };

    let pem = general_purpose::STANDARD
        .decode(key_b64.trim())
        .context("GITHUB_APP_PRIVATE_KEY: invalid base64")
        .map_err(AppError::Internal)?;

    let encoding_key = EncodingKey::from_rsa_pem(&pem)
        .context("GITHUB_APP_PRIVATE_KEY: invalid RSA PEM")
        .map_err(AppError::Internal)?;

    let now = chrono::Utc::now().timestamp();
    let claims = AppClaims {
        iat: now - 60,
        exp: now + 540, // 9 min (max 10 min allowed by GitHub)
        iss: app_id,
    };

    let token = encode(&Header::new(Algorithm::RS256), &claims, &encoding_key)
        .context("Error signing JWT RS256")
        .map_err(AppError::Internal)?;

    Ok(Some(token))
}

/// Gets an installation access token for the GitHub App.
///
/// The token is valid for 1 hour. This function generates it on demand —
/// it is not cached because most operations are not high frequency.
///
/// [Fix #18] Receives `http: &reqwest::Client` from AppState — never create
/// `Client::new()` per request (new TCP pool per call, exhausts descriptors).
pub async fn get_installation_access_token(
    state: &AppState,
    installation_id: &str,
) -> Result<Option<String>, AppError> {
    let Some(app_jwt) = build_app_jwt(state).await? else {
        tracing::warn!("GitHub App not configured — GITHUB_APP_ID or GITHUB_APP_PRIVATE_KEY absent");
        return Ok(None);
    };

    let resp = state
        .http
        .post(format!(
            "https://api.github.com/app/installations/{installation_id}/access_tokens"
        ))
        .header("Authorization", format!("Bearer {app_jwt}"))
        .header("Accept", "application/vnd.github+json")
        .header("X-GitHub-Api-Version", "2022-11-28")
        .header("User-Agent", "plane-api-rust/0.1")
        .send()
        .await
        .context("Error contacting GitHub API")
        .map_err(AppError::Internal)?;

    if resp.status().is_success() {
        let body: serde_json::Value = resp
            .json()
            .await
            .context("GitHub API: response is not valid JSON")
            .map_err(AppError::Internal)?;

        Ok(body["token"].as_str().map(String::from))
    } else {
        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();
        tracing::warn!(
            installation_id,
            %status,
            body = %text,
            "GitHub App installation token request failed"
        );
        Ok(None)
    }
}
